### ISSUE TITLE
Integrated terminal needs multi-line input support

### ISSUE THREAD

**Issue by ChrisLynchHPE** [PowerShell/vscode-powershell#569]
### System Details

- Operating system name and version: Windows 10.15058 (Insiders Fast Ring)
- VS Code version: 1.10.2
- PowerShell extension version: 0.10
- Output from `$PSVersionTable`:

```
Name                           Value
----                           -----
PSVersion                      5.1.15058.0
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15058.0
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### Issue Description

I have a Hashtable in my PowerShell module, and the new terminal in 0.10 does not work correctly like it used to.  Paste the following into the Terminal:

```PowerShell
$MyHT = @{
   key1 = 'value1';
   key2 = 'value2
}
```
And you will get the following output:

```PowerShell
PS C:\Users\chris> $MyHT = @{
The hash literal was incomplete.
At line:0 char:0
PS C:\Users\chris>    key1 = 'value1';
key1 : The term 'key1' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included,
verify that the path is correct and try again.
At line:1 char:4
+    key1 = 'value1';
+    ~~~~
    + CategoryInfo          : ObjectNotFound: (key1:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

PS C:\Users\chris>    key2 = 'value2
The string is missing the terminator: '.
At line:0 char:0
PS C:\Users\chris> }
```

Not only does one get the exceptions above, but the terminal is extremely slow.  I don't know if it's because of the loaded PSM1 file I have (which is more than 80k lines, and 2.2MB in size).

### Attached Logs

When I enable plugin verbose logging, `powershell.exe` will terminate and I can't then reproduce the issue:

```
powershell.exe started --
    pid: 940
    exe: C:\WINDOWS\Sysnative\WindowsPowerShell\v1.0\powershell.exe
    args: C:\Users\chris\.vscode\extensions\ms-vscode.PowerShell-0.10.0\scripts\Start-EditorServices.ps1 -EditorServicesVersion '0.10.0' -HostName 'Visual Studio Code Host' -HostProfileId 'Microsoft.VSCode' -HostVersion '0.10.0' -BundledModulesPath 'C:\Users\chris\.vscode\extensions\ms-vscode.PowerShell-0.10.0\modules' -EnableConsoleRepl -LogLevel 'Verbose' -LogPath 'C:\Users\chris\.vscode\extensions\ms-vscode.PowerShell-0.10.0\logs\1489692702-86da6b33-6d62-41a2-a912-eb66f3cd355a1489692658699\EditorServices.log' -SessionDetailsPath 'C:\Users\chris\.vscode\extensions\ms-vscode.PowerShell-0.10.0\sessions\PSES-VSCode-12712' 



powershell.exe terminated or terminal UI was closed

Could not start language service: 
Timed out waiting for session file to appear.
```
(the log file generated contains the text I pasted above.)

**Comment by daviwil** [comment]
Hey Chris, this is happening because I haven't added support for multi-line input.  I'll try to get this in for 0.10.2.

**Comment by ChrisLynchHPE** [comment]
Ah, okay.  Also, any idea on why the powershell process would be so slow on my Surface Pro 3 Core i5/256 system?  If I terminate the integrated console or PowerShell session in VSCode, overall performance dramatically improves.  I don't have any similar perf issues with PowerShell outside of VSCode.

**Comment by Jaykul** [comment]
Multi-line badly needed. Also, text navigation in input, like `Ctrl+Left` or `Ctrl+Shift+Left` or `Ctrl+A` or `Ctrl+Backspace` ;-)

**Comment by daviwil** [comment]
I was trying to decide whether I should add text navigation or let PSReadline take care of it once I get that  integration working.  Since that may take a little while, maybe it's better to get some basic navigation hotkeys working?

**Comment by bgelens** [comment]
Yes please 👍 Heavy user of ```ctrl+left``` and ```ctrl+right``` myself

**Comment by Jaykul** [comment]
If PSReadLine integration was easy, that's obviously very nice ;-) But basic multi-line and text editor cursor movement keys would go a long way toward making a long wait comfortable ;-)

**Comment by benpiper** [comment]
Please make it so!

**Comment by el-chazmo** [comment]
This still doesn't work in v1.18.1 Is this in the timeline?

**Comment by tylerl0706** [comment]
@el-chazmo thanks for your patience! This will be fixed when the Integrated Terminal supports PSReadline. @SeeminglyScience has been working hard on adding it. I don't have a timeline but he might be able to supply one. Or to supply an update.

**Comment by Jaykul** [comment]
+Label AintGonnaDoIt☹️

Can we agree that after supposedly trying to get PSReadLine integrated for over a year, it's _**long** past time_ to improve _your own_ input?  I mean: how many people actually type in that console, the way it is?

Nobody on my team will use the interactive debugger, because working in that terminal is so horrible.

**Comment by SeeminglyScience** [comment]
@Jaykul It's mostly done, PR is PowerShell/PowerShellEditorServices#672 if you want to try it out 😀

**Comment by CollinChaffin** [comment]
The Posh masses really need to STOP posting on every blog how vscode is in ANY way the "best" Posh editor/IDE to be using (as of today).  I commend the effort but I cannot fathom anyone coming from a real Posh background using IDEs like Pstudio/Primalforms/conemu with npp/posh split-screened or heck - how about just plain old Notepad++ with powershell.exe opened below it for that matter calling vscode a step-up.
 
I'm trying to understand this part of over a full YEAR of "waiting for psReadline" to work.  Have you taken a look at something like ConEmu, who's darned source code is openly posted for you to review, and since that **_single individual developer_** DID manage to quickly get FAR BEYOND psreadline usability, and in fact successfully process every single Posh (and heck, you can literally flip the script and get it to swallow up vscode itself now for that matter as shown below) and working pretty much error-free 100% of the time for most of us.....then why is it again that vscode with the entire MS DEV at it's disposal can't get even close to the conemu equivalent terminal emulator for only the Posh/PoshCore hosting/emulation?

The issue is when everywhere I look vscode is being pushed as "the" IDE with "terminal" for Posh DEV right now, yet I cannot even so much as paste a simple multi-line splatted param or hashtable into the only viable Posh "terminal" in vscode without every line vomiting red because the interpreter has no concept of what a line continuation even is...or as Joel said total lack of basic navigation.....despite the fact that by now it really should be, it is unfortunately IMO still not ready for serious Posh development usage and there is no way I can come close to moving coding efforts into vscode....but after 30+ yrs of being a MS evangelist I guess a boy can still dream! :wink:   

<br>

_If it helps, here is vscode being pulled into Conemu as a child docked terminal with full injection and ANSI support: (So why can't powershell.exe or pwsh.exe be integrated into vscode the exact same way proven to work for years in conemu?  It's using WINAPI and injection all fully supported methods last I checked)_

![84a7f14e-16e5-11e6-933b-96f5f5f9ccc2](https://user-images.githubusercontent.com/9301483/46566086-e7f94200-c8dd-11e8-9818-b2219eaf0626.png)

**Comment by SeeminglyScience** [comment]
@CollinChaffin The difficult part wasn't getting PSReadLine to work in VSCode.

The challenge was getting PSReadLine to work in the same PowerShell `runspace` as editor features like intellisense and debugging.   Both need to be in the main pipeline thread, and finding a way to get them both to coexist in the same thread without completely rewriting them wasn't easy.  As far as I'm aware there hasn't been any other editors that have implemented support for PSReadLine in a way that was integrated with editor features.  The ISE is the closest I know of, but instead of supporting PSReadLine they wrote their own ReadLine implementation.  Granted, that's still better than what is currently released for VSCode, but full PSReadLine support is ideal.

All that said, full support for PSReadLine is implemented.  The latest preview build is available in [this comment](https://github.com/PowerShell/vscode-powershell/issues/535#issuecomment-418944258) if you'd like to help test.

**Comment by sdsalyer** [comment]
Just popping in to cast my vote for multi-line support in the extension's "PowerShell Integrated Console". It works fine in the standard console opend via `ctrl+``, so it's frustrating when you open a ps1 file and suddenly it "stops working".

**Comment by TylerLeonhardt** [comment]
@sdalyer the "PowerShell Preview" extension in the marketplace has PSReadLine support and multi-line support.

Please check out our blog post here:
https://blogs.msdn.microsoft.com/powershell/2019/01/23/announcing-the-powershell-preview-extension-in-vscode/

**Comment by SydneyhSmith** [comment]
Closing as PSReadLine is now available in the PowerShell Preview extension, tracking that PSReadLine is not available in the main PowerShell Extension here #1793
